packages/components/src/components/Calculator/index.tsx

```diff
@@ -0,0 +1,152 @@
+};
+
+export const Calculator: FC<Props> = ({
+  value: calculator,
+  environment,
+  settings,
+  renderValue,
+  valueWithContext,
+}) => {
+  const { path } = valueWithContext.context;
+  const { getSettings, dispatch } = useViewerContext();
+  const [, forceUpdate] = useReducer((x) => x + 1, 0);
+
+  const itemSettings = getSettings({ path, defaults: undefined });
+  const calculatorState = itemSettings.calculator;
+
+  const init = async () => {
+    await initialize({
+      dispatch,
+      path,
+      calculator,
+      environment,
+    });
+    forceUpdate();
+  };
+
+  useEffect(() => {
+    !calculatorState && init();
+  }, []);
+
+  const handleChange =
+    (name: string) =>
+    async (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+      const newCode = e.target.value;
+
+      calculatorState &&
+        (await updateCode({
+          dispatch,
+          path,
+          environment: environment,
+          state: calculatorState,
+          name,
+          code: newCode,
+          forceUpdate,
+          calculator,
+        }));
+    };
+
+  const showSqValue = (
+    item: result<SqValue, SqError>,
+    settings: PlaygroundSettings
+  ) => {
+    if (item.ok) {
+      const value = item.value;
+      if (valueHasContext(value)) {
+        return renderValue(value, settings);
+      } else {
+        return value.toString();
+      }
+    } else {
+      return (
+        <div className="text-sm text-red-800 text-opacity-70">
+          {item.value.toString()}
+        </div>
+      );
+    }
+  };
+
+  const fieldShowSettings: PlaygroundSettings = {
+    ...settings,
+    distributionChartSettings: {
+      ...settings.distributionChartSettings,
+      showSummary: false,
+    },
+    chartHeight: 30,
+  };
+
+  const resultSettings: PlaygroundSettings = {
+    ...settings,
+    chartHeight: 200,
+  };
+
+  return (
+    <div className="relative space-y-4">
+      {calculator.description && (
+        <ReactMarkdown className={"prose text-sm text-slate-800 bg-opacity-60"}>
+          {calculator.description}
+        </ReactMarkdown>
+      )}
+
+      {calculatorState &&
+        calculator.rows.map((row) => {
+          const { name, description } = row;
+          const { value, code } = calculatorState!.fields[name];
+          const result = value;
+          const resultHasInterestingError = result && !result.ok && code !== "";
+          return (
+            <div key={name} className="flex flex-col max-w-lg">
+              <div className="text-sm font-semibold text-slate-800">{name}</div>
+              {description && (
+                <div className="text-sm  text-slate-600">{description}</div>
+              )}
+              <div className="flex-grow">
+                <input
+                  value={code || ""}
+                  onChange={handleChange(name)}
+                  placeholder={`Enter code for ${name}`}
+                  className="my-2 p-2 border rounded w-full"
```

Comment:
I can see why you wouldn't want to bother with react-hook-form, but you could still use [StyledInput](https://quri-ui.vercel.app/?path=/docs/ui-forms-styledinput--docs) for consistency.